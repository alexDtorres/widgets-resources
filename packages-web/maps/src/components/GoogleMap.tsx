import { createElement, CSSProperties, ReactElement, useEffect, useRef, useState } from "react";
import classNames from "classnames";

import googleApiWrapper from "./GoogleApi";
import { Alert } from "@widgets-resources/piw-utils";
import { getDimensions } from "../utils";
import { Marker, SharedProps } from "../../typings";

export interface GoogleMapsProps extends SharedProps {
    scriptsLoaded?: boolean;
    validationMessage?: string;
    mapsToken?: string;
    className?: string;
    mapStyles?: string;
    divStyles?: CSSProperties;

    optionStreetView: boolean;
    mapTypeControl: boolean;
    fullScreenControl: boolean;
    rotateControl: boolean;
}

export const GoogleMap = (props: GoogleMapsProps): ReactElement => {
    const map = useRef<google.maps.Map>();
    const googleMapsRef = useRef<HTMLDivElement>(null);
    const defaultCenterLocation: google.maps.LatLngLiteral = { lat: 51.9066346, lng: 4.4861703 };
    let bounds!: google.maps.LatLngBounds;

    const [markers, setMarkers] = useState<google.maps.Marker[]>([]); //Used to manage and remove markers from the map
    const [validationMessage, setValidationMessage] = useState(props.validationMessage);

    useEffect(() => {
        if (props.scriptsLoaded) {
            createUpdateMap();
        }
    }, []);

    useEffect(() => {
        createUpdateMap();
    }, [props.locations, props.currentLocation]);

    const createUpdateMap = (): void => {
        const mapOptions: google.maps.MapOptions = {
            zoom: props.zoomLevel,
            zoomControl: props.optionZoomControl,
            scrollwheel: props.optionScroll,
            draggable: props.optionDrag,
            streetViewControl: props.optionStreetView,
            mapTypeControl: props.mapTypeControl,
            fullscreenControl: props.fullScreenControl,
            rotateControl: props.rotateControl,
            minZoom: 1,
            maxZoom: 20,
            styles: getMapStyles()
        };
        if (googleMapsRef.current && !map.current) {
            map.current = new google.maps.Map(googleMapsRef.current, mapOptions);
        } else if (map.current) {
            map.current.setOptions(mapOptions);
        }
        addMarkers();
    };

    const addMarkers = (): void => {
        if (props.locations && props.locations.length > 0) {
            markers.forEach(marker => marker.setMap(null));
            setMarkers([]);
            bounds = new google.maps.LatLngBounds();
            setMarkers(
                props.locations.reduce<google.maps.Marker[]>((markerArray, currentLocation) => {
                    const marker = addMarker(currentLocation);
                    if (marker) {
                        markerArray.push(marker);
                    }
                    return markerArray;
                }, [])
            );
        }
        addCurrentLocation();
        updateCamera();
    };

    const addCurrentLocation = () => {
        if (props.currentLocation) {
            const currentLocationMarker = addMarker(props.currentLocation);
            if (currentLocationMarker) {
                markers.push(currentLocationMarker);
            }
        }
    };

    const addMarker = (marker: Marker): google.maps.Marker | undefined => {
        if (map.current) {
            bounds.extend({
                lat: marker.latitude,
                lng: marker.longitude
            });
            const mapMarker = new google.maps.Marker({
                position: {
                    lat: marker.latitude,
                    lng: marker.longitude
                },
                icon: marker.url,
                title: marker.title
            });
            if (marker.title) {
                const infoContent = document.createElement("span");
                infoContent.innerHTML = marker.title || "";
                if (marker.onClick) {
                    infoContent.style.cursor = "pointer";
                    infoContent.onclick = marker.onClick;
                }
                const infoWindow = new google.maps.InfoWindow({
                    content: infoContent
                });
                mapMarker.addListener("click", () => {
                    infoWindow.open(map.current, mapMarker);
                });
            } else {
                if (marker.onClick) {
                    mapMarker.addListener("click", () => {
                        marker.onClick!();
                    });
                } else {
                    mapMarker.setClickable(false);
                }
            }
            mapMarker.setMap(map.current);
            return mapMarker;
        }
        return undefined;
    };

    const updateCamera = (): void => {
        const { zoomLevel, autoZoom } = props;
        setTimeout(() => {
            if (bounds && map.current) {
                try {
                    if (!autoZoom) {
                        map.current.setCenter({
                            lat: markers[0].getPosition()?.lat() ?? defaultCenterLocation.lat,
                            lng: markers[0].getPosition()?.lng() ?? defaultCenterLocation.lng
                        });
                        map.current.setZoom(zoomLevel);
                    } else {
                        map.current.fitBounds(bounds);
                    }
                } catch (error) {
                    setValidationMessage(`Invalid map bounds ${error.message}`);
                }
            }
        }, 0);
    };

    const getMapStyles = (): google.maps.MapTypeStyle[] => {
        if (props.mapStyles && props.mapStyles.trim()) {
            try {
                return JSON.parse(props.mapStyles);
            } catch (error) {
                setValidationMessage(`invalid Map styles, ${error.message}`);
            }
        }

        return [
            {
                featureType: "poi",
                elementType: "labels",
                stylers: [{ visibility: "off" }]
            }
        ];
    };

    return (
        <div
            className={classNames("widget-maps", props.className)}
            style={{ ...props.divStyles, ...getDimensions(props) }}
        >
            {validationMessage && (
                <Alert bootstrapStyle="danger" className="widget-google-maps-alert">
                    {validationMessage}
                </Alert>
            )}
            <div className="widget-google-maps-wrapper">
                <div className="widget-google-maps" ref={googleMapsRef} />
            </div>
        </div>
    );
};

export default googleApiWrapper("https://maps.googleapis.com/maps/api/js?key=")(GoogleMap);
