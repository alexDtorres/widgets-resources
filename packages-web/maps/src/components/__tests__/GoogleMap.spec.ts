import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";
import { GoogleMap, GoogleMapsProps } from "../GoogleMap";
import { initialize } from "@googlemaps/jest-mocks";

describe("Google maps", () => {
    const defaultProps: GoogleMapsProps = {
        autoZoom: true,
        height: 75,
        heightUnit: "pixels",
        scriptsLoaded: true,
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        zoomLevel: 10,
        width: 50,
        widthUnit: "percentage",
        mapStyles: "",
        style: {},
        currentLocation: undefined,
        optionStreetView: false,
        locations: [],
        className: "",
        fullScreenControl: false,
        mapsToken: "",
        mapTypeControl: false,
        rotateControl: false,
        showCurrentLocation: false,
        validationMessage: ""
    };

    beforeEach(() => {
        initialize();
    });

    const renderGoogleMap = (props: GoogleMapsProps): ShallowWrapper<GoogleMapsProps, any> =>
        shallow(createElement(GoogleMap, props));

    it("renders a map with right structure", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with pixels renders structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "pixels",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of width and height units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of parent units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    afterAll(() => {
        window.google = undefined;
    });
});
