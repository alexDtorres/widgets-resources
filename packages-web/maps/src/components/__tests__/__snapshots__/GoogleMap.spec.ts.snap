// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google maps renders a map with percentage of parent units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75%",
      "width": "50%",
    }
  }
>
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
</div>
`;

exports[`Google maps renders a map with percentage of width and height units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "auto",
      "paddingBottom": "37.5%",
      "width": "50%",
    }
  }
>
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
</div>
`;

exports[`Google maps renders a map with pixels renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50px",
    }
  }
>
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
</div>
`;

exports[`Google maps renders a map with right structure 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "37.5px",
      "width": "50px",
    }
  }
>
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
</div>
`;
